File: VotingSystem.API/Features/UserProfile/Services/UserProfileService.cs
using VotingSystem.API.Features.UserProfile.DTOs;
using VotingSystem.Common.ResponseModel;
using VotingSystem.Infrastructure.Repositories.Interfaces;

namespace VotingSystem.API.Features.UserProfile.Services
{
    public class UserProfileService : IUserProfileService
    {
        private readonly IUserRepository _userRepository;

        public UserProfileService(IUserRepository userRepository)
        {
            _userRepository = userRepository;
        }

        public async Task<ApiResponse<UserResDTO>> GetUserDetail(string userId)
        {
            var user = await _userRepository.GetById(userId);
            if (user is null)
            {
                return ApiResponse<UserResDTO>.Failed(null, "No user found");
            }

            #region response model mapping
            var mappedModel = new UserResDTO
            {
                FullName = user.FullName,
                Username = user.UserName ?? "",
                Email = user.Email ?? ""
            };
            #endregion

            return ApiResponse<UserResDTO>.Success(mappedModel, "User detail loaded");
        }

        public async Task<ApiResponse<UserResDTO>> UpdateUser(string userId, UpdateUserDTO dto)
        {
            var user = await _userRepository.GetById(userId);
            if (user is null)
            {
                return ApiResponse<UserResDTO>.Failed(null, "No user found");
            }

            if (!string.IsNullOrEmpty(dto.FullName))
            {
                user.FullName = dto.FullName;
            }
            if (!string.IsNullOrEmpty(dto.Username))
            {
                user.UserName = dto.Username;
            }
            if (!string.IsNullOrEmpty(dto.Email))
            {
                user.Email = dto.Email;
            }

            var result = await _userRepository.Update(user);
            if (result is null)
            {
                return ApiResponse<UserResDTO>.Failed(null, "Failed to update user");
            }

            #region response model mapping
            var mappedModel = new UserResDTO
            {
                FullName = result.FullName,
                Username = result.UserName ?? "",
                Email = result.Email ?? ""
            };
            #endregion

            return ApiResponse<UserResDTO>.Success(mappedModel, "User detail loaded");
        }
    }
}

File: VotingSystem.XFrontend/src/app/services/userProfile/user-profile.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../../../environment';
import { Observable } from 'rxjs';
import { UpdateUserDto } from '../../models/DTOs/update-user.dto';
import { HttpClient } from '@angular/common/http';
import { ApiResponse } from '../../models/api-response.model';

@Injectable({
  providedIn: 'root',
})
export class UserProfileService {
  baseUrl = environment.apiBaseUrl;

  constructor(private http: HttpClient) {}

  getUser(): Observable<ApiResponse<UpdateUserDto>> {
    return this.http.get<ApiResponse<UpdateUserDto>>(`${this.baseUrl}/User/get-user`);
  }

  updateUser(data: UpdateUserDto): Observable<ApiResponse<UpdateUserDto>> {
    return this.http.post<ApiResponse<UpdateUserDto>>(`${this.baseUrl}/User/update-user`, data);
  }
}
